File: src/discovery/indeed.py
import requests
from bs4 import BeautifulSoup
from typing import Dict, List, Optional

from src.utils.logger import error, log
from src.config import JOB_KEYWORDS, JOB_LOCATION, MAX_AGE_DAYS, REMOTE_ONLY, USER_AGENT


def build_indeed_url(keywords: str, location: str, max_age_days: int, remote_only: bool) -> str:
    params = {
        "q": keywords,
        "l": location,
        "fromage": str(max_age_days),
        "sort": "date",
        "radius": "0",
        "limit": "50",
    }
    if remote_only:
        params["remotejob"] = "1"
    query = "&".join(f"{key}={requests.utils.quote(value)}" for key, value in params.items())
    return f"https://www.indeed.com/jobs?{query}"


def parse_posted_age(text: str) -> Optional[str]:
    if not text:
        return None
    value = text.lower().strip()
    if "just posted" in value or "today" in value:
        return None
    return None


def fetch_indeed_jobs() -> List[Dict]:
    url = build_indeed_url(JOB_KEYWORDS, JOB_LOCATION, MAX_AGE_DAYS, REMOTE_ONLY)
    log(f"Fetching Indeed jobs from {url}")

    headers = {
        "User-Agent": USER_AGENT,
        "Accept-Language": "en-US,en;q=0.9",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    }

    response = requests.get(url, headers=headers, timeout=20)
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser")
    jobs = []

    selectors = [".job_seen_beacon", ".result", ".tapItem"]
    for selector in selectors:
        for card in soup.select(selector):
            link = card.select_one("a[data-jk], a[title], a[href*='/rc/clk'], a[href*='/company/']")
            if not link:
                continue
            job_path = link.get("href", "")
            if not job_path:
                continue
            job_url = job_path if job_path.startswith("http") else f"https://www.indeed.com{job_path}"
            title_tag = card.select_one("h2.jobTitle span, h2.title span, .jobTitle")
            company_tag = card.select_one("span.companyName, .company, .companyName")
            location_tag = card.select_one("div.companyLocation, .location")
            snippet_tag = card.select_one("div.job-snippet, .summary")
            posted_tag = card.select_one("span.date, .date")

            jobs.append(
                {
                    "source": "Indeed",
                    "job_title": title_tag.get_text(strip=True) if title_tag else "Data Engineering",
                    "company": company_tag.get_text(strip=True) if company_tag else "Unknown",
                    "location": location_tag.get_text(strip=True) if location_tag else JOB_LOCATION,
                    "remote": REMOTE_ONLY or "remote" in (location_tag.get_text(strip=True).lower() if location_tag else ""),
                    "posted_date": parse_posted_age(posted_tag.get_text(strip=True)) if posted_tag else None,
                    "url": job_url,
                    "description": snippet_tag.get_text(strip=True) if snippet_tag else "",
                    "requirements": None,
                    "salary": None,
                    "id": f"indeed-{requests.utils.quote(job_url)}",
                    "raw_payload": {
                        "posted_text": posted_tag.get_text(strip=True) if posted_tag else "",
                        "snippet": snippet_tag.get_text(strip=True) if snippet_tag else ""
                    },
                }
            )

    log(f"Indeed discovered {len(jobs)} jobs")
    return jobs

File: src/discovery/indeed.js
import axios from "axios";
import * as cheerio from "cheerio";
import { log, error } from "../utils/logger.js";

const USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";

function parsePostedAge(text) {
  const cleaned = text?.toLowerCase().trim() ?? "";
  if (!cleaned) {
    return null;
  }
  if (cleaned.includes("just posted") || cleaned.includes("today")) {
    return new Date().toISOString();
  }
  const match = cleaned.match(/(\d+)\+?\s*day/);
  if (match) {
    const daysAgo = Number(match[1]);
    const date = new Date();
    date.setDate(date.getDate() - daysAgo);
    return date.toISOString();
  }
  return null;
}

function buildIndeedUrl({ keywords, location, maxAgeDays, remoteOnly }) {
  const params = new URLSearchParams();
  params.set("q", keywords);
  params.set("l", location);
  params.set("fromage", String(maxAgeDays));
  params.set("sort", "date");
  params.set("radius", "0");
  if (remoteOnly) {
    params.set("remotejob", "1");
  }
  params.set("limit", "50");
  return `https://www.indeed.com/jobs?${params.toString()}`;
}

async function fetchJobDetail(url) {
  try {
    const response = await axios.get(url, {
      headers: {
        "User-Agent": USER_AGENT,
        "Accept-Language": "en-US,en;q=0.9",
        Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
      },
      timeout: 20000
    });
    const $ = cheerio.load(response.data);
    const jobDescription = $("#jobDescriptionText").text().trim() || $(".jobsearch-jobDescriptionText").text().trim();
    return jobDescription || null;
  } catch (err) {
    error(`Indeed detail fetch failed for ${url}: ${err.message}`);
    return null;
  }
}

export async function fetchIndeedJobs({ keywords, location, remoteOnly, maxAgeDays }) {
  const url = buildIndeedUrl({ keywords, location, remoteOnly, maxAgeDays });
  log(`Fetching Indeed jobs from ${url}`);

  const response = await axios.get(url, {
    headers: {
      "User-Agent": USER_AGENT,
      "Accept-Language": "en-US,en;q=0.9",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      "Referer": "https://www.indeed.com/"
    },
    timeout: 20000
  });

  const $ = cheerio.load(response.data);
  const jobs = [];

  $(".job_seen_beacon, .result, .tapItem").each((index, element) => {
    const node = $(element);
    const link = node.find("a[data-jk], a[title], a[href*='/rc/clk'], a[href*='/company/']").first();
    const jobPath = link.attr("href") || "";
    if (!jobPath) {
      return;
    }

    const url = jobPath.startsWith("http") ? jobPath : `https://www.indeed.com${jobPath}`;
    const title = node.find("h2.jobTitle span, h2.title span, .jobTitle").first().text().trim();
    const company = node.find("span.companyName, .company, .companyName").first().text().trim();
    const locationText = node.find("div.companyLocation, .location").first().text().trim();
    const snippet = node.find("div.job-snippet, .summary").text().trim();
    const postedText = node.find("span.date, .date").first().text().trim();
    const postedDate = parsePostedAge(postedText) || new Date().toISOString();

    jobs.push({
      source: "Indeed",
      job_title: title || "Data Engineering",
      company: company || "Unknown",
      location: locationText || "United States",
      remote: remoteOnly || /remote/i.test(locationText),
      posted_date: postedDate,
      url,
      description: snippet,
      requirements: null,
      salary: null,
      id: `indeed-${encodeURIComponent(url)}`,
      raw_payload: {
        posted_text: postedText,
        snippet
      }
    });
  });

  log(`Indeed discovered ${jobs.length} jobs`);
  return jobs.slice(0, 50);
}
